### PR TITLE
[https://nvbugs/5451205][feat] Add cuBLASLt NVFP4 GEMM backend support

### PR DESCRIPTION
**Pull request by Wong4j**
## Summary by CodeRabbit

- New Features
  - Optional FP4 GEMM acceleration via cuBLASLt, with runtime availability detection.
  - New FP4 scaled matrix-multiply op exposed to PyTorch `(torch.ops.trtllm.cublas_fp4_scaled_mm)`.
  - Linear layer option to use the cuBLASLt FP4 path; BF16 output supported.
  - Preserves existing paths when cuBLASLt is unavailable or the option is disabled.
  - (Update Oct 22) Further support for autotune allows selecting the fastest kernel from the algorithms returned by the cublaslt heuristic.

- Performance
  - Potential speedups for FP4 matrix multiplications on supported GPUs.

- Tests
  - Added unit and performance tests validating the cuBLASLt FP4 path across shapes and dtypes.

## Description

- New Features
  - Added an optional nvfp4 block-scaled gemm path. Supports heuristic algorithm selection and bf16 output.
  - Added a corresponding argument in the linear layer to enable it; disabled by default. When disabled, the original behavior is maintained.
  - Added unit tests comparing with the existing cutlass implementation, with consistent results.

## Test Coverage

op UT:
pytest -s -o log_cli=true tests/unittest/_torch/thop/parallel/test_fp4_linear.py -k "test_fp4_linear_cublaslt"
image

model UT:
pytest -s -o log_cli=true "tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4"
image

## Perf
**All tests use autotune to traverse and select the fastest kernel.**
| Shape (M×N×K) | CUTLASS (μs) | cuBLASLt (μs) | Speedup |
|:--------------|-------------:|--------------:|--------:|
| 8192×8192×1024 | 57.34 | 60.77 | 0.94x |
| 8192×8192×2048 | 92.83 | 83.62 | 1.11x |
| 8192×8192×4096 | 151.17 | 140.06 | 1.08x |
| 8192×8192×8192 | 267.49 | 250.62 | 1.07x |
| 8192×8192×16384 | 484.86 | 459.97 | 1.05x |

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by rosenrodt — COMMENTED** [review]
I left some comments inline. Thank you!

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py [hunk 1]
```diff
@@ -390,6 +390,159 @@ def nvfp4_gemm_perf_test(
...
+    alpha_ref = 1.0 / (w_sf_global * x_sf_global)
+    alpha_tensor = torch.tensor(alpha_ref, dtype=torch.float32).cuda()
+    beta_tensor = torch.tensor(0.0, dtype=torch.float32).cuda()
+
+    if test_ref:
+        with nvtx.annotate(
+                f"cublaslt tune, m={SEQ_LEN}, k={HIDDEN_SIZE}, n={OUTPUT_SIZE}",
+                color="blue"):
+            with torch.inference_mode(), autotune():
+                output_cublaslt = torch.ops.trtllm.cublas_fp4_scaled_mm(
```
[the remark this answers is not in the document]
Comment (Wong4j) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: I have added autotune support.

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py [hunk 2]
```diff
@@ -390,6 +390,153 @@ def nvfp4_gemm_perf_test(
...
+
+    # Calculate correct alpha scaling factor
+    alpha_ref = 1.0 / (w_sf_global * x_sf_global)
+    alpha_tensor = torch.tensor(alpha_ref, dtype=torch.float32).cuda()
+
+    if test_ref:
+        with nvtx.annotate(
+                f"cublaslt tune, m={SEQ_LEN}, k={HIDDEN_SIZE}, n={OUTPUT_SIZE}",
+                color="blue"):
+            with torch.inference_mode(), autotune():
```
Comment (rosenrodt) [inline tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: If test_ref is false, then `nvfp4_gemm_cublaslt` is never run with `autotune()` context. Is that intended?

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 3]
```diff
@@ -431,8 +431,69 @@ def forward(
...
+            mat2,
+            mat1_scale,
+            mat2_scale,
+            alpha,
+            self.to_userbuffers,
+            tactic,
+        )
+        return result
+
+    def set_best_tactic(self, inputs: List[torch.Tensor], best_tactic: int):
```
Comment (rosenrodt) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Is `set_best_tactic()` API being used somewhere?

File: cpp/tensorrt_llm/common/cublasMMWrapper.h [hunk 4]
```diff
@@ -83,6 +83,13 @@ class CublasMMWrapper
         int const lda, void const* B, int const ldb, void* C, int const ldc, float f_alpha, float f_beta,
         cublasLtMatmulAlgo_t const& algo, bool hasAlgo, bool usingCublasLt);
 
+#ifdef ENABLE_CUBLASLT_FP4_GEMM
+    /********************** FP4 GEMMs **********************/
+    void Fp4Gemm(cublasOperation_t transa, cublasOperation_t transb, int const m, int const n, int const k,
```
[the remark this answers is not in the document]
Comment (rosenrodt) [reply cpp/tensorrt_llm/common/cublasMMWrapper.h]: Thanks

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py [hunk 5]
```diff
@@ -390,6 +390,159 @@ def nvfp4_gemm_perf_test(
...
+                mat_a=x_fp4,
+                mat_b=w_fp4,
+                scale_a=x_sf_block,
+                scale_b=w_sf_block,
+                alpha=alpha_tensor,
+                beta=beta_tensor,
+                out_dtype=dtype)
+
+    if test_ref:
+        torch.testing.assert_close(output_cublaslt,
```
[the remark this answers is not in the document]
Comment (Wong4j) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: Yes, you’re right. This perf_test was written based on an existing test in this file. In practice, performance still needs to be manually compared using nvtx and nsys profiling. Repeated checks are unnecessary and have been removed.

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py (hunk 2 above)
Comment (Wong4j) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: removed

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py (hunk 3 above)
Comment (rosenrodt) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Thanks

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py (hunk 2 above)
Comment (rosenrodt) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: I see `autotune()` is removed from the test. But we probably want to test it. Is that unintended?

**Review by hyukn — APPROVED** [review]
LGTM. I suggest enabling DEBUG logger to use UT to check that tuning happens as expected and AutoTuner cache is filled with correct shapes.

**Review by rosenrodt — APPROVED** [review]
PR looks good and shows perf improvement. Thanks a lot and good work!

**Review by MartinMarciniszyn — APPROVED** [review]

**Review by hlu1 — APPROVED** [review]

**Review by HuiGao-NV — APPROVED** [review]
LGTM
